dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-reports-reconciliation-summary.controller/helpers.ts

```diff
@@ -0,0 +1,169 @@
+import { getAllUtilisationDataForReport } from '../../../../services/repositories/utilisation-data-repo';
+import {
+  getUtilisationReportDetailsByBankIdMonthAndYear,
+  getOpenReportsBeforeReportPeriodForBankId,
+} from '../../../../services/repositories/utilisation-reports-repo';
+import {
+  getPreviousReportPeriodStart,
+  getReportPeriodStartForSubmissionMonth,
+  getSubmissionMonthForReportPeriodStart,
+  isEqualReportPeriodStart,
+} from '../../../../utils/report-period';
+
+type UtilisationReportForSubmissionMonth = {
+  submissionMonth: IsoMonthStamp;
+  report: UtilisationReport | null;
+};
+
+type SummaryItemForSubmissionMonth = {
+  submissionMonth: IsoMonthStamp;
+  item: UtilisationReportReconciliationSummaryItem;
+};
+
+const mapToSummaryItem = async (bank: Bank, report: UtilisationReport | null): Promise<UtilisationReportReconciliationSummaryItem> => {
+  // TODO FN-???? - Use actual REPORT_NOT_RECEIVED UtilisationReport when added by job. Throw errors when null.
+  if (!report) {
+    return {
+      bank: {
+        id: bank.id,
+        name: bank.name,
+      },
+      status: UTILISATION_REPORT_RECONCILIATION_STATUS.REPORT_NOT_RECEIVED,
+    };
+  }
+
+  const reportData = await getAllUtilisationDataForReport(report);
+
+  // TODO FN-???? - status to be added to report data to allow us to calculate how
+  //  many facilities are left to reconcile
+  const reportedFeesLeftToReconcile = reportData.length;
+
+  return {
+    reportId: report._id,
+    bank: {
+      id: bank.id,
+      name: bank.name,
+    },
+    status: report.status,
+    dateUploaded: report.dateUploaded,
+    totalFeesReported: reportData.length,
+    reportedFeesLeftToReconcile,
+    isPlaceholderReport: report.azureFileInfo === null,
+  };
+};
+
+const mapToSummaryItemForSubmissionMonth = async (
+  bank: Bank,
+  { submissionMonth, report }: UtilisationReportForSubmissionMonth,
+): Promise<SummaryItemForSubmissionMonth> => ({
+  submissionMonth,
+  item: await mapToSummaryItem(bank, report),
+});
+
+const generateMissingReports = ({ from, to }: { from: IsoMonthStamp; to: IsoMonthStamp }): UtilisationReportForSubmissionMonth[] => {
+  const missingSubmissionMonths = eachIsoMonthOfInterval(from, to, {
+    exclusive: true,
+  });
+
+  return missingSubmissionMonths.map((submissionMonth) => ({
+    submissionMonth,
+    report: null,
+  }));
+};
+
+const addNotReceivedReportsAndMapToSubmissionMonth = (
+  reports: UtilisationReport[],
+  currentSubmissionMonth: IsoMonthStamp,
+): UtilisationReportForSubmissionMonth[] => {
+  const currentReportPeriodStart = getReportPeriodStartForSubmissionMonth(currentSubmissionMonth);
+  const previousReportPeriodStart = getPreviousReportPeriodStart(currentReportPeriodStart);
+
+  const reportsOrderedByReportPeriodStartAscending = orderBy(reports, ['year', 'month'], ['asc', 'asc']);
+
+  return reportsOrderedByReportPeriodStartAscending.reduce((reportsWithSubmissionMonth, report, index) => {
```

Comment:
I don't think the `reduce` is appropriate here. You only perform the check for `reportsAreMissing` at the last index, so I think this would be easier to read if written like this:
```
const updatedReportsWithSubmissionMonth = reportsOrderedByReportPeriodStartAscending.map((report) => {
  const reportPeriodStart: ReportPeriodStart = { month: report.month, year: report.year };
  const submissionMonth = getSubmissionMonthForReportPeriodStart(reportPeriodStart);
  return { submissionMonth, report };
});

const mostRecentOpenReport = reportsOrderedByReportPeriodStartAscending.at(-1);
const mostRecentOpenReportPeriodStart: ReportPeriodStart = { month: mostRecentOpenReport.month, year: mostRecentOpenReport.year };

if (isEqualReportPeriodStart(mostRecentOpenReportPeriod, previousReportPeriod)) {
  return updatedReportsWithSubmissionMonth;
}

const mostRecentOpenReportSubmissionMonth = updatedReportsWithSubmissionMonth.at(-1).submissionMonth;
const missingReports = generateMissingReports({
  from: mostRecentOpenReportSubmissionMonth,
  to: currentSubmissionMonth,
});
return [...updatedReportsWithSubmissionMonth, ...missingReports];
```
Haven't tested this, but I think something like this is clearer than `reduce` in this case

Reply:
This is much clearer, thanks